### PR TITLE
Add dynamic configuration for Namespaces [HZ-2576]

### PR DESCRIPTION
**Pull request by JamesHazelcast**
This PR focuses on the changes required for the dynamic configuration of Namespaces.
[Namespaces feature TDD for reference](https://gist.github.com/JamesHazelcast/f25007a4e0594c491ce867c816b616a7).

_This PR is 1 of 5 pertaining to the Namespace feature addition. The other 4 PRs can be found below:_
- https://github.com/hazelcast/hazelcast/pull/26144
- https://github.com/hazelcast/hazelcast/pull/26146
- https://github.com/hazelcast/hazelcast/pull/26147
- https://github.com/hazelcast/hazelcast/pull/26148

_Due to the Namespaces feature being split into 5 pieces for easier reviewing, it is highly recommended that you clone the [UCD feature development branch](https://github.com/vbekiaris/hazelcast/tree/enhancements/5.4.0/ucd) for testing purposes, as these individual PRs are likely non-functional standalone._

### CODE REVIEW

**Review by kwart — COMMENTED** [review]
Looks good permission-wise.

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java [hunk 1]
```diff
@@ -389,4 +393,19 @@ default UUID updateAsync() {
...
      */
     Map<String, WanReplicationConfig> getWanReplicationConfigs();
+
+    /**
+     * Finds existing namespace configuration by name.
+     *
+     * @param name name of the configuration
+     * @return Namespace configuration or {@code null} when the requested configuration does not exist
+     */
+    NamespaceConfig findNamespaceConfig(String name);
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java:403]: `findNamespaceConfig` and `getNamespaceConfigs` below seem unused in the project. They should be removed. In particular, `findNamespaceConfig` is irrelevant for namespace configs since we don't apply the lookup-by-pattern config matching style that we apply for data structure configuration.

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java [hunk 2]
```diff
@@ -344,13 +367,37 @@ public void registerConfigLocally(IdentifiedDataSerializable newConfig, ConfigCh
...
+     *
+     * @param newConfig       Configuration to register.
+     * @throws UnsupportedOperationException when given configuration type is not supported
+     */
+    @SuppressWarnings("checkstyle:methodlength")
+    public void deregisterConfigLocally(IdentifiedDataSerializable newConfig) {
+        if (newConfig instanceof NamespaceConfig) {
+            NamespaceConfig config = (NamespaceConfig) newConfig;
+            if (namespaceConfigs.remove(config.getName(), newConfig)) {
+                nodeEngine.getNamespaceService().removeNamespaceConfig(config);
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java:393]: The call to `namespaceService.removeNamespaceConfig` should be always executed, since we are supposed to allow for removal of statically configured namespaces. The current code with the condition above will not work as expected in this scenario:

- start member with statically configured namespace `ns1`
- client invokes `hz.getConfig().getNamespacesConfig().removeNamespace("ns1")`
- since there is no entry in dynamically configured `namespaceConfigs` map maintained in `ClusterWideConfigurationService`, namespace is not removed from `NamespaceService`

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java [hunk 3]
```diff
@@ -152,6 +167,13 @@ String generate(Config config, boolean isMask) {
         networkConfigYamlGenerator(root, config);
         advancedNetworkConfigYamlGenerator(root, config);
         dataConnectionYamlGenerator(root, config);
+        namespacesConfigGenerator(root, config);
+        metricsConfigGenerator(root, config);
+        splitBrainProtectionConfigsGenerator(root, config);
+
+        // Reset maskSensitiveFields to default
+        DynamicConfigYamlGenerator.maskSensitiveFields = DEFAULT_MASK_SENSITIVE_FIELDS;
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java:175]: Why is this required?

**Review by vbekiaris — CHANGES_REQUESTED** [review]
great job overall!

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java (hunk 1 above)
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java:403]: `getNamespaceConfigs` makes sense imo since we're supporting NamespaceConfigs in dynamic configs - agreed on the `findNamespaceConfig` method though 👍

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java (hunk 3 above)
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java:175]: This method changes the static field `maskSensitiveFields` to the passed variable when executed, so it made sense to me to return the state to how it was prior once we were finished - happy to remove.

Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java:175]: I think you are potentially fixing a bug here. If some code called `generate` with `mask == true` (didn't find any in production but still..), then any subsequent invocation of any other method in this class would generate masked values - including the config persistence feature which would result in a config file persisted with masked license key -> won't work on next restart.

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java (hunk 1 above)
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ConfigurationService.java:403]: Addressed in https://github.com/vbekiaris/hazelcast/commit/cafc0048ba946af66a094d8832bc1dcfb59eed5d

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java (hunk 2 above)
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java:393]: Addressed in https://github.com/vbekiaris/hazelcast/commit/cafc0048ba946af66a094d8832bc1dcfb59eed5d

File: hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java (hunk 3 above)
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java:175]: Restored in https://github.com/vbekiaris/hazelcast/commit/9f686c70bada9c576b0d8d8b149cf3d429c9ab80

**Review by vbekiaris — APPROVED** [review]
all my comments have been addressed, thank you @JamesHazelcast

**Review by frant-hartm — APPROVED** [review]

**Comment by JamesHazelcast** [comment]
Closing this PR due to the creation of the final mono-repo PR: https://github.com/hazelcast/hazelcast-mono/pull/217 🎉
